### PR TITLE
TRRWriter, use ts.data['step'] if it exists

### LINKED ISSUE #3350: TRRWriter doesn't use ts.data['step'] when copying TRR file
I'm not entirely sure if this is a bug because I'm not entirely sure what `ts.data['step']` is _supposed_ to be. But in any case, I think copying a file by writing the entire thing to a new file should maintain all of the data in the original file, which in this case it's not.

## Expected behavior ##

Writing an entire trajectory to a new file should maintain all of the data.

## Actual behavior ##

The `ts.data['step']` information is lost when copying a TRR file by writing with MDAnalysis

## Code to reproduce the behavior ##

``` python
import MDAnalysis as mda
from MDAnalysis.tests.datafiles import TPR_xvf, TRR_xvf

u = mda.Universe(TPR_xvf, TRR_xvf)

for ts in u.trajectory:
    print (ts.data['step'])
0
25000
50000

with mda.Writer('copied_file.trr', n_atoms=u.trajectory.n_atoms) as W:
    for ts in u.trajectory:
        W.write(u)

copied_u = mda.Universe(TPR_xvf, 'copied_file.trr')

# now has different steps
for ts in copied_u.trajectory:
    print (ts.data['step'])
0
1
2

```

## Current version of MDAnalysis ##

- Which version are you using? (run `python -c "import MDAnalysis as mda; print(mda.__version__)"`)
2.0.0-dev0
- Which version of Python (`python -V`)?
Python 3.8.5
- Which operating system?
Windows 10 Home

### PR DESCRIPTION
**Pull request by ianmkenney**
If ts.data['step'] exists, use this for the output
step value. Otherwise, use ts.frame, consistent
with previous behavior.

Fixes #3350 

Changes made in this Pull Request:
 - use `ts.data.get('step', ts.frame)` instead of `ts.frame` for the step variable

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Review by hmacdope — CHANGES_REQUESTED** [review]
Thanks for the PR @ianmkenney!
 Would you be able to add a test to cover the writing behaviour?

**Comment by ianmkenney** [comment]
@hmacdope Test added.

**Review by richardjgowers — CHANGES_REQUESTED** [review]
Ch ch ch ch CHANGELOG

File: testsuite/MDAnalysisTests/coordinates/test_xdr.py
Comment (hmacdope) [inline testsuite/MDAnalysisTests/coordinates/test_xdr.py]: Could you please use an fstring rather than % format here :)

File: testsuite/MDAnalysisTests/coordinates/test_xdr.py
Comment (hmacdope) [inline testsuite/MDAnalysisTests/coordinates/test_xdr.py]: This will also pass if both are 'None' or empty. Can we also check that 'ts.data' contains the data we expect? 

Follow up question is there data in ts.data for the original file?

**Review by hmacdope — CHANGES_REQUESTED** [review]
Just a nitpick and a question :)

**Comment by hmacdope** [comment]
Also need to address PEP8.

File: testsuite/MDAnalysisTests/coordinates/test_xdr.py
Comment (ianmkenney) [reply testsuite/MDAnalysisTests/coordinates/test_xdr.py]: `ts.data` contains this information. I was just patching the behavior described in #3350 with the solution described by https://github.com/MDAnalysis/mdanalysis/issues/3350#issuecomment-864844782 which indicates that `ts.data['step']` might not exist in some cases. I can check the contents of the test file, if that's what you mean.

```
{'time': 0.0, 'step': 0, 'lambda': 0.0, 'dt': 100.00000762939453}
{'time': 100.00000762939453, 'step': 50000, 'lambda': 0.0, 'dt': 100.00000762939453}
{'time': 200.00001525878906, 'step': 100000, 'lambda': 0.0, 'dt': 100.00000762939453}
{'time': 300.0, 'step': 150000, 'lambda': 0.0, 'dt': 100.00000762939453}
{'time': 400.0000305175781, 'step': 200000, 'lambda': 0.0, 'dt': 100.00000762939453}
{'time': 500.0000305175781, 'step': 250000, 'lambda': 0.0, 'dt': 100.00000762939453}
{'time': 600.0, 'step': 300000, 'lambda': 0.0, 'dt': 100.00000762939453}
{'time': 700.0000610351562, 'step': 350000, 'lambda': 0.0, 'dt': 100.00000762939453}
{'time': 800.0000610351562, 'step': 400000, 'lambda': 0.0, 'dt': 100.00000762939453}
{'time': 900.0000610351562, 'step': 450000, 'lambda': 0.0, 'dt': 100.00000762939453}
```

**Comment by ianmkenney** [comment]
@hmacdope how offensive do you find this to be? I can also make this change in other parts of the file if you want.

```
n_atoms = universe.atoms.n_atoms                                                                                                                                                    
dt = universe.trajectory.dt                                                                                                                                                         
                                                                                                                                                                          
with Writer(outfile, n_atoms, dt) as W:                                                                                                                                             
    for ts in universe.trajectory:                                                                                                                                                  
    W.write(universe)  
```

File: testsuite/MDAnalysisTests/coordinates/test_xdr.py
Comment (ianmkenney) [reply testsuite/MDAnalysisTests/coordinates/test_xdr.py]: Sure

File: testsuite/MDAnalysisTests/coordinates/test_xdr.py
Comment (hmacdope) [reply testsuite/MDAnalysisTests/coordinates/test_xdr.py]: Sorry I should have been clearer, my bad, was just checking that `ts.data` has something in it in the test file. All sorted.

**Comment by hmacdope** [comment]
You mean as a replacement for:

```python
    def test_data(self, universe, Writer, outfile):
        with Writer(outfile, universe.atoms.n_atoms, dt=universe.trajectory.dt) as W:
            for ts in universe.trajectory:
                W.write(universe)
```
I kind of like the way that it is now, as everything is nice and encapsulated in its owning object (no unnecessary variables), but it is your code and if you think its clearer the second way then go for it.

 Other people may also have an opinion @lilyminium?

Perhaps leave the rest of the file the way it is. We generally avoid cosmetic changes for their own sake, as often original writers have made something look the way it does for a reason (often one that you may not expect), so always safer to leave something in its original state.

**Comment by ianmkenney** [comment]
@hmacdope I have the same opinion as you regarding the extra variables. I'd also rather contribute something consistent. That said, I've also updated the CHANGELOG.

File: testsuite/MDAnalysisTests/coordinates/test_xdr.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/coordinates/test_xdr.py]: There's no cost here between checking that the contents exist or that they match a certain value. Can we do the latter for these instead?

File: testsuite/MDAnalysisTests/coordinates/test_xdr.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/coordinates/test_xdr.py]: are we likely to be comparing floats here? if so we should use an almost_equal for the data components that are floats?

File: package/MDAnalysis/coordinates/TRR.py
Comment (IAlibay) [inline package/MDAnalysis/coordinates/TRR.py:116]: Can you add a versionchanged entry please?

Can you maybe update the blurb on line 48? Whilst you're here, can you fix the typo on line 44 ("velocoties").

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple of comments - mostly updating the docs.

**Review by richardjgowers — APPROVED** [review]
I'm happy once @IAlibay is happy

File: testsuite/MDAnalysisTests/coordinates/test_xdr.py
Comment (ianmkenney) [reply testsuite/MDAnalysisTests/coordinates/test_xdr.py]: Checking with the values found in the first frame rather than `not None`

File: testsuite/MDAnalysisTests/coordinates/test_xdr.py
Comment (ianmkenney) [reply testsuite/MDAnalysisTests/coordinates/test_xdr.py]: using `assert_equal` on the dictionary itself

File: package/MDAnalysis/coordinates/TRR.py
Comment (ianmkenney) [reply package/MDAnalysis/coordinates/TRR.py:116]: Fixed two typos ('velocoties'), added versionchanged for 2.1.0, and appended to the docstring following line 48 as an example.

File: testsuite/MDAnalysisTests/coordinates/test_xdr.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/coordinates/test_xdr.py]: I think you want assert_almost_equal for floats

**Review by IAlibay — CHANGES_REQUESTED** [review]
Just the one final thing

File: testsuite/MDAnalysisTests/coordinates/test_xdr.py
Comment (hmacdope) [reply testsuite/MDAnalysisTests/coordinates/test_xdr.py]: Note that `assert_almost_equal` is being phased out in favour of `assert_allclose` but makes no difference here AFAIK. Im happy either way.

Comment (ianmkenney) [reply testsuite/MDAnalysisTests/coordinates/test_xdr.py]: Just going to use `assert_allclose`

**Review by IAlibay — APPROVED** [review]

**Comment by IAlibay** [comment]
thanks for dealing with this @ianmkenney

**Comment by IAlibay** [comment]
@hmacdope you have a blocking review - since you've commented since I'll wait until tomorrow morning for your approval, after that I'll squash-merge (and dismiss blocking reviews if necessary).

**Review by hmacdope — APPROVED** [review]
LGTM! Thanks @ianmkenney

**Pull request merged**
